File: frontend/dreamboard/src/app/models/image-gen-models.ts
/***************************************************************************
 *
 *  Copyright 2025 Google Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  Note that these code samples being shared are not official Google
 *  products and are not formally supported.
 *
 ***************************************************************************/

export interface Image {
  id: string;
  name: string;
  gcsUri: string;
  signedUri: string;
  gcsFusePath: string;
  mimeType: string;

}

export interface ImageReference extends Image {
  referenceId?: number;
  referenceType: string; // Image reference flag.
  referenceSubType: string; // Secondary flag (e.g. Subject Person, Subject animal).
  description?: string;
  maskMode?: string;
  maskDilation?: number;
  segmentationClasses?: number[];
  enableControlImageComputation?: boolean;
}

export interface ImageGenerationSettings {
  prompt: string;
  numImages: number;
  aspectRatio?: string;
  outputMimeType?: string;
  compressionQuality?: number;
  language?: string;
  safetyFilterLevel?: string;
  personGeneration?: string;
  seed?: number;
  negativePrompt?: string;
  selectedImageForVideo?: Image; // Image used to generate the video
  referenceImages?: ImageReference[]; // Image used to generate new images with AI, if selected, can also be used to generate the video
  generatedImages: Image[]; // Contains AI generated images and reference images
  useReferenceImageForImage?: boolean;
  editMode?: string;
}

/* Models for backend interactions */

export interface ImageCreativeDirection {
  aspect_ratio: string;
  model?: string;
  number_of_images: number;
  output_mime_type: string;
  person_generation: string;
  safety_filter_level: string;
  output_gcs_uri?: string[];
  language: string;
  output_compression_quality: number;
  negative_prompt?: string;
  enhance_prompt?: boolean;
  seed?: number;
  // add_watermark: boolean;
}

export interface ReferenceImageCard {
  id: string;
}

/* Models for backend interactions */

export interface ImageItem {
  id?: string;
  name: string;
  gcs_uri: string;
  signed_uri: string;
  gcs_fuse_path: string;
  mime_type: string;
}

export interface ImageReferenceItem extends ImageItem {
  reference_id: number;
  reference_type: string; // Image reference flag.
  reference_subtype?: string; // Secondary flag (e.g. Subject Person, Subject animal).
  description?: string;
  mask_mode?: string;
  mask_dilation?: number;
  segmentation_classes?: number[];
  enable_control_image_computation?: boolean;
}

export interface ImageSceneRequest {
  scene_num: number;
  img_prompt: string;
  image_uri?: string[];
  scene_id?: string[];
  creative_dir?: ImageCreativeDirection;
  image_content_type?: string;
  reference_images?: ImageReferenceItem[];
  use_reference_image_for_image?: boolean;
  edit_mode?: string;
}

export interface ImageSceneCreativeDirection {
  numImages: number;
  resolution: string;
  aspectRatio: string;
}

export interface SceneSegment {
  default_images: string[];
  imageScene: string;
  imagePrompt: string;
  imageSelected: string;
  creativeDirection: ImageSceneCreativeDirection;
}

export interface ImageSceneSegments {
  scenes: ImageSceneRequest[];
}

export interface ImageGenerationRequest {
  scenes: ImageSceneRequest[];
}

export interface ImageGenerationResponse {
  scene_ids: string;
  segment_number: number;
  done: boolean;
  operation_name: string;
  execution_message: string;
  images: ImageItem[];
}

File: backend/app/models/image/image_gen_models.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Image Generation data classes.

This file contains data models (dataclasses) used across the application
to represent various entities related to image generation and file handling,
particularly in the context of the Imagen API.
"""

import enum
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class Image:
  """
  Represents a single image, typically as a response from an Imagen API
  generation or as an input reference.

  Attributes:
      name: An optional name for the image.
      gcs_uri: The Google Cloud Storage (GCS) URI where the image is stored.
      mime_type: The MIME type of the image (e.g., 'image/jpeg', 'image/png').
      signed_uri: A pre-signed URL for temporary public access to the GCS URI.
      gcs_fuse_path: The FUSE path if the GCS bucket is mounted locally.
      image_bytes: The raw bytes of the image, if available in memory.
  """

  name: Optional[str] | None = None
  gcs_uri: Optional[str] | None = None
  mime_type: Optional[str] | None = None
  signed_uri: Optional[str] | None = None
  gcs_fuse_path: Optional[str] | None = None
  image_bytes: Optional[bytes] | None = None


class IMAGE_REFERENCE_TYPES(enum.Enum):
  """
  Defines the different types of reference images supported by the Imagen API
  for image editing operations.
  """

  RAW = "raw"
  MASK = "mask"
  SUBJECT = "subject"
  STYLE = "style"
  CONTROLLED = "controlled"


@dataclass
class ImageReference(Image):
  """
  Extends the `Image` class to include attributes specific to reference images
  used in Imagen API's image editing capabilities.

  Attributes:
      reference_id: A unique identifier for this specific reference.
      reference_type: The broad category of the reference (e.g., "raw",
                      "mask"). Uses `IMAGE_REFERENCE_TYPES` values.
      reference_subtype: A more specific classification within the type
                         (e.g., "Person", "Animal" for 'subject' type).
      description: A text description providing more context for the reference
                   (e.g., for 'style' or 'subject' references).
      mask_mode: For 'mask' references, specifies how the mask should be
                 interpreted (e.g., "apply", "ignore").
      mask_dilation: For 'mask' references, an optional value to expand or
                     contract the mask area.
      segmentation_classes: For 'mask' references, a list of integer IDs
                            representing object classes to be masked.
      enable_control_image_computation: For 'controlled' references, a flag
                                        to enable specific control image
                                        processing.
  """

  # Shared attributes.
  reference_id: Optional[int] = None
  reference_type: Optional[str] | None = None

  # Partially shared attributes.
  # Secondary flag (e.g. Subject Person, Subject animal).
  reference_subtype: Optional[str] | None = None
  description: Optional[str] | None = None

  # Mask Reference specific attributes.
  mask_mode: Optional[str] | None = None
  mask_dilation: Optional[float] | None = None
  segmentation_classes: Optional[List[int]] = None

  # Control Reference specific attributes.
  enable_control_image_computation: bool = False


@dataclass
class ImageGenerationResponse:
  """
  Represents the structured response received after an image generation or
  editing request to the Imagen API.

  Attributes:
      scene_ids: A string identifier for the overall generation batch.
      segment_number: An integer indicating the specific segment or part
                      of a larger generation process.
      done: A boolean flag indicating if the operation is complete.
      operation_name: The name of the asynchronous operation (e.g., for
                      polling its status).
      execution_message: Any message or status detail about the execution.
      images: A list of `Image` objects representing the generated images.
  """

  scene_ids: str
  segment_number: int
  done: bool
  operation_name: str
  execution_message: str
  images: list[Image]
